src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs

```diff
@@ -0,0 +1,187 @@
+                        if (_cancellationToken.IsCancellationRequested)
+                        {
+                            _source.SetException(new OperationCanceledException(_cancellationToken));
+                            return;
+                        }
+
+                        notifyWhenUnblocked = ThreadPool.NotifyThreadBlocked();
+                        switch (_operation)
+                        {
+                            case Operation.Read:
+                                Memory<byte> writableSingleSegment = Unsafe.As<ReadOnlyMemory<byte>, Memory<byte>>(ref _singleSegment);
+                                result = RandomAccess.ReadAtOffset(_fileHandle, writableSingleSegment.Span, _fileOffset);
+                                break;
+                            case Operation.Write:
+                                result = RandomAccess.WriteAtOffset(_fileHandle, _singleSegment.Span, _fileOffset);
+                                break;
+                            case Operation.ReadScatter:
+                                Debug.Assert(_multiSegmentCollection != null && _multiSegmentCollection is IReadOnlyList<Memory<byte>>);
+                                result = RandomAccess.ReadScatterAtOffset(_fileHandle, (IReadOnlyList<Memory<byte>>) _multiSegmentCollection, _fileOffset);
+                                break;
+                            case Operation.WriteGather:
+                                Debug.Assert(_multiSegmentCollection != null && _multiSegmentCollection is IReadOnlyList<ReadOnlyMemory<byte>>);
+                                result = RandomAccess.WriteGatherAtOffset(_fileHandle, (IReadOnlyList<ReadOnlyMemory<byte>>)_multiSegmentCollection, _fileOffset);
+                                break;
+                        }
+                    } finally
+                    {
+                        if (notifyWhenUnblocked)
+                            ThreadPool.NotifyThreadUnblocked();
+                        _operation = Operation.None;
+                        _fileOffset = 0;
+                        _cancellationToken = default;
+                        _singleSegment = default;
+                        _multiSegmentCollection = null;
+                    }
+                } catch (Exception e)
+                {
+                    _source.SetException(e);
+                }
+                _source.SetResult(result);
+            }
+
+            public ValueTask<int> QueueRead(Memory<byte> buffer, long fileOffset, CancellationToken cancellationToken)
+            {
+                Debug.Assert(_operation == Operation.None, "An operation was queued before the previous one's completion.");
+
+                _operation = Operation.Read;
+                _singleSegment = buffer;
+                _fileOffset = fileOffset;
+                _cancellationToken = cancellationToken;
+                ThreadPool.UnsafeQueueUserWorkItem(this, false);
```

Comment:
Should I set `preferLocal` to `true`? We can assume that the `ValueTask` is going to be awaited soon and the same thread to become quickly free to pick another work item.